### ISSUE TITLE
dns: srv update causes type mismatch in stats output when port is >32767

### ISSUE THREAD

**Issue by lukastribus** [haproxy/haproxy#103]
Mustafa Yildirim reported this issue on discourse:

https://discourse.haproxy.org/t/problem-when-using-dynamic-service-discovery-via-dns-and-srv-records/3818

## Output of `haproxy -vv` and `uname -a`

```
lukas@dev:~/haproxy$ ./haproxy -vv
HA-Proxy version 2.0-dev3-9c8800-42 2019/05/20 - https://haproxy.org/
Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_GETADDRINFO=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO -OPENSSL -LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 -ZLIB -SLZ +CPU_AFFINITY -TFO -NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without compression support (neither USE_ZLIB nor USE_SLZ are set).
Compression algorithms supported : identity("identity")
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): yes

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE
              h2 : mode=HTTP       side=FE
       <default> : mode=HTX        side=FE|BE
       <default> : mode=TCP|HTTP   side=FE|BE

Available services : none

Available filters :
        [SPOE] spoe
        [COMP] compression
        [CACHE] cache
        [TRACE] trace

lukas@dev:~/haproxy$ uname -a
Linux dev 4.4.0-146-generic #172-Ubuntu SMP Wed Apr 3 09:00:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
lukas@dev:~/haproxy$

```

## What's the configuration?

```
global
 log /dev/log    local0
 log /dev/log    local1 notice
 log 10.0.0.4 syslog debug
 maxconn 2000
 stats socket /var/run/haproxy-1.sock uid 0 gid 0 mode 0666 level admin process 1

defaults
 log     global
 mode http
 timeout connect 5000
 timeout client  50000
 timeout server  50000
 timeout http-keep-alive 60s
 default-server init-addr none

resolvers localdns
 nameserver dns1 8.8.8.8:53
 hold valid 60s

frontend f
 bind :80
 default_backend b

backend b
 server srv1 _testport32768._tcp.ltri.eu resolvers localdns resolve-prefer ipv4 resolve-opts allow-dup-ip
```

## Steps to reproduce the behavior

1. start haproxy with config above, using server port `32768` (srv record in the example config is valid, must be higher than 32767)
2. check port with "show servers state" on the admin socket
3. check port with "show stat" on the admin socket

## Actual behavior

Instead of `32768`, "show servers state" shows `4294934528` and "show stat" shows `-32768`:

```
lukas@dev:~$ echo "show servers state" | socat stdio unix-connect:/var/run/haproxy-1.sock
1
# be_id be_name srv_id srv_name srv_addr srv_op_state srv_admin_state srv_uweight srv_iweight srv_time_since_last_change srv_check_status srv_check_result srv_check_health srv_check_state srv_agent_state bk_f_forced_id srv_f_forced_id srv_fqdn srv_port srvrecord
4 b 1 srv1 127.0.0.1 2 0 1 1 35 1 0 2 0 0 0 0 localhost 4294934528 _testport32768._tcp.ltri.eu

lukas@dev:~$ echo "show stat" typed | socat stdio unix-connect:/var/run/haproxy-1.sock | grep addr
S.4.1.73.addr.1:CGS:str:127.0.0.1:-32768
lukas@dev:~$

```

## Expected behavior

Show the correct port in all situations.

## Do you have any idea what may have caused this?

Likely there's a mismatch between what the DNS stack writes (uint16 ??), what "show stat" (int?) and what "show servers state" (unsigned int?) print.

Only happens with ports provided by SRV records.

## Do you have an idea how to solve the issue?

I don't.

**Comment by wtarreau** [comment]
I have absolutely no idea how the SRV update works, but it's obvious here that there is a short signed int converted to a signed int. I'll have a quick look but I think @bedis would be more fluent in this area!

**Comment by wtarreau** [comment]
Care to retry with this ?
```
diff --git a/include/types/dns.h b/include/types/dns.h
index 81cd6d260..5a60c0708 100644
--- a/include/types/dns.h
+++ b/include/types/dns.h
@@ -146,7 +146,7 @@ struct dns_answer_item {
        int32_t         ttl;                         /* response TTL */
        int16_t         priority;                    /* SRV type priority */
        uint16_t        weight;                      /* SRV type weight */
-       int16_t         port;                        /* SRV type port */
+       uint16_t        port;                        /* SRV type port */
        uint16_t        data_len;                    /* number of bytes in target below */
        struct sockaddr address;                     /* IPv4 or IPv6, network format */
        char            target[DNS_MAX_NAME_SIZE+1]; /* Response data: SRV or CNAME type target */
```
Also I suspect that some of the other fields like TTL or priority suffer from the same problem but have less impact here.

**Comment by lukastribus** [comment]
Confirmed, this fixes the bug.

**Comment by wtarreau** [comment]
Cool, thank you Lukas! I've just pushed the patch (commit d1f56c9a0).

**Comment by lukastribus** [comment]
Christopher backported the fix to 1.9 and 1.8.
